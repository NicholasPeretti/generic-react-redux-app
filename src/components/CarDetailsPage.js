import React from 'react'
import PropTypes from 'prop-types'
import CarPropType from '../utils/propTypes/car'
import styled from 'styled-components'
import Style from '../style'
import Title from './Title'
import CarTitle from './CarTitle'
import CarShortDescription from './CarShortDescription'
import Text from './Text'
import SubTitle from './SubTitle'
import Button from './Button'
import BorderedBox from './BorderedBox'

const Wrapper = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;
`

const Container = styled.div`
  width: 800px;
  display: flex;
`

const Column = styled.div`
  display: flex;
  flex-direction: column;
`

const RowReverse = styled.div`
  display: flex;
  flex-direction: row-reverse;
`

const Space = styled.div`
  width: 100%;
  height: ${Style.spacing.padding3};
`

const CarDetailsPage = ({ car, saveCar, unsaveCar, isCarSaved }) => (
  <Wrapper>
    <Container>
      <Column style={{ width: '60%' }}>
        <Column>
          <Title>
            <CarTitle {...car} />
          </Title>
          <Space />
          <SubTitle>
            <CarShortDescription {...car} />
          </SubTitle>
          <Space />
          <Text>
            This car is currently available and can be delivered as soon as
            tomorrow morning. Please be aware that delivery times shown in this
            page are not definitive and may change due to bad weather
            conditions.
          </Text>
        </Column>
      </Column>
      <Column style={{ width: '40%', marginLeft: Style.spacing.padding3 }}>
        <BorderedBox>
          <Column>
            <Text>
              If you like this car, click the button and save it in your
              collection of favourite items.
            </Text>
            <RowReverse>
              <Button onClick={isCarSaved ? unsaveCar : saveCar}>
                {isCarSaved ? 'Unsave' : 'Save'}
              </Button>
            </RowReverse>
          </Column>
        </BorderedBox>
      </Column>
    </Container>
  </Wrapper>
)

CarDetailsPage.propTypes = {
  car: CarPropType.isRequired,
  saveCar: PropTypes.func.isRequired,
  unsaveCar: PropTypes.func.isRequired,
  isCarSaved: PropTypes.bool.isRequired
}

export default CarDetailsPage
