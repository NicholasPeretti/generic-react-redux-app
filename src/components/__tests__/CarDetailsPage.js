import React from 'react'
import { shallow, mount } from 'enzyme'
import PresentationalTest from '../../utils/tests/presentational'
import CarDetailsPage from '../CarDetailsPage'
import { CarsList } from '../../utils/tests/data'
const MOCK_CAR = CarsList[0]

const props = {
  car: MOCK_CAR,
  isCarSaved: false,
  saveCar: jest.fn(),
  unsaveCar: jest.fn()
}
const carSavedProps = Object.assign({}, props, {
  isCarSaved: true
})

describe('CarDetailsPage', () => {
  PresentationalTest(CarDetailsPage, props)

  test('Should match snapshot when the car is not saved', () => {
    const wrapper = shallow(<CarDetailsPage {...props}/>)
    expect(wrapper.html()).toMatchSnapshot()
  })
  test('Should match snapshot when the car is saved', () => {
    const wrapper = shallow(<CarDetailsPage {...carSavedProps}/>)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Should call saveCar when the proper button is clicked', () => {
    const wrapper = mount(<CarDetailsPage {...props}/>)
    wrapper.find('button').simulate('click')
    expect(props.saveCar).toHaveBeenCalledTimes(1)
  })

  test('Should call unsaveCar when the proper button is clicked', () => {
    const wrapper = mount(<CarDetailsPage {...carSavedProps}/>)
    wrapper.find('button').simulate('click')
    expect(carSavedProps.unsaveCar).toHaveBeenCalledTimes(1)
  })
})
