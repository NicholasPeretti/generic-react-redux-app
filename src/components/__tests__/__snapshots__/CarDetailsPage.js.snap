// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarDetailsPage Should match snapshot when the car is not saved 1`] = `"<div class=\\"sc-EHOje dUxNKn\\"><div class=\\"sc-bZQynM fmYkVX\\"><div style=\\"width:60%\\" class=\\"sc-gzVnrw kvhoeX\\"><div class=\\"sc-gzVnrw kvhoeX\\"><h1 class=\\"sc-bdVaJa cyYLae\\">BMW X1</h1><div class=\\"sc-dnqmqq hZDOWW\\"></div><h3 class=\\"sc-htpNat ldqxTF\\">Stock # 10079 - 178994 km - Diesel - black</h3><div class=\\"sc-dnqmqq hZDOWW\\"></div><span class=\\"sc-bwzfXH isOLYK\\">This car is currently available and can be delivered as soon as tomorrow morning. Please be aware that delivery times shown in this page are not definitive and may change due to bad weather conditions.</span></div></div><div style=\\"width:40%;margin-left:24px\\" class=\\"sc-gzVnrw kvhoeX\\"><div class=\\"sc-ifAKCX javJuV\\"><div class=\\"sc-gzVnrw kvhoeX\\"><span class=\\"sc-bwzfXH isOLYK\\">If you like this car, click the button and save it in your collection of favourite items.</span><div class=\\"sc-htoDjs bhGQoG\\"><button class=\\"sc-bxivhb kyvnHz\\"><span class=\\"sc-bwzfXH isOLYK\\">Save</span></button></div></div></div></div></div></div>"`;

exports[`CarDetailsPage Should match snapshot when the car is saved 1`] = `"<div class=\\"sc-EHOje dUxNKn\\"><div class=\\"sc-bZQynM fmYkVX\\"><div style=\\"width:60%\\" class=\\"sc-gzVnrw kvhoeX\\"><div class=\\"sc-gzVnrw kvhoeX\\"><h1 class=\\"sc-bdVaJa cyYLae\\">BMW X1</h1><div class=\\"sc-dnqmqq hZDOWW\\"></div><h3 class=\\"sc-htpNat ldqxTF\\">Stock # 10079 - 178994 km - Diesel - black</h3><div class=\\"sc-dnqmqq hZDOWW\\"></div><span class=\\"sc-bwzfXH isOLYK\\">This car is currently available and can be delivered as soon as tomorrow morning. Please be aware that delivery times shown in this page are not definitive and may change due to bad weather conditions.</span></div></div><div style=\\"width:40%;margin-left:24px\\" class=\\"sc-gzVnrw kvhoeX\\"><div class=\\"sc-ifAKCX javJuV\\"><div class=\\"sc-gzVnrw kvhoeX\\"><span class=\\"sc-bwzfXH isOLYK\\">If you like this car, click the button and save it in your collection of favourite items.</span><div class=\\"sc-htoDjs bhGQoG\\"><button class=\\"sc-bxivhb kyvnHz\\"><span class=\\"sc-bwzfXH isOLYK\\">Unsave</span></button></div></div></div></div></div></div>"`;
